### PR TITLE
Adding Bulk Add to driver API

### PR DESCRIPTION
**Pull request by kellrott**
This adds a BulkAdd method to the internal driver interface (and methods to each of the drivers). This allows for Badger driver to stream all the writes into a single bulk transaction (rather than having to do one batch at a time) and as a result, it does over 10x faster on large loads.

### CODE REVIEW

File: kvgraph/graph.go [hunk 1]
```diff
@@ -74,6 +57,57 @@ func (kgdb *KVInterfaceGDB) AddVertex(vertices []*gripql.Vertex) error {
...
+	if err := tx.Set(key, value); err != nil {
+		return fmt.Errorf("AddVertex Error %s", err)
+	}
+	if err := idx.AddDocTx(tx, vertex.Gid, doc); err != nil {
+		return fmt.Errorf("AddVertex Error %s", err)
+	}
+	return nil
+}
+
+func insertEdge(tx kvi.KVBulkWrite, idx *kvindex.KVIndex, graph string, edge *gripql.Edge) error {
```
Comment (adamstruck) [inline kvgraph/graph.go]: AddEdge should be using this method.

File: server/api.go [hunk 2]
```diff
@@ -169,133 +171,66 @@ func (server *GripServer) addEdge(ctx context.Context, elem *gripql.GraphElement
...
+					graph, err := server.db.Graph(element.Graph)
+					if err != nil {
+						log.WithFields(log.Fields{"error": err}).Error("BulkAdd: graph not found")
+						loopErr = err
+					}
+					elementStream = make(chan *gripql.GraphElement, 100)
+					wg.Add(1)
+					go func() {
+						log.Printf("Streaming to %s", element.Graph)
+						graph.BulkAdd(elementStream)
```
Comment (adamstruck) [inline server/api.go]: BulkAdd error needs to be handled.

File: server/api.go [hunk 3]
```diff
@@ -169,133 +171,66 @@ func (server *GripServer) addEdge(ctx context.Context, elem *gripql.GraphElement
...
+						close(elementStream)
+					}
+					graph, err := server.db.Graph(element.Graph)
+					if err != nil {
+						log.WithFields(log.Fields{"error": err}).Error("BulkAdd: graph not found")
+						loopErr = err
+					}
+					elementStream = make(chan *gripql.GraphElement, 100)
+					wg.Add(1)
+					go func() {
```
Comment (adamstruck) [inline server/api.go]: This go routine should be outside the main while loop.

File: util/insert.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	if edgeBufSize != 0 {
+		return &graphElementArray{edges: make([]*gripql.Edge, 0, edgeBufSize)}
+	}
+	return nil
+}
+
+// SteamBatch a stream of inputs and loads them into the graph
+// This function assumes incoming stream is GraphElemnts from a single graph
+func SteamBatch(stream <-chan *gripql.GraphElement, vertexAdd func([]*gripql.Vertex) error, edgeAdd func([]*gripql.Edge) error) error {
```
Comment (adamstruck) [inline util/insert.go]: Should we consider having batch size as an argument?

File: kvgraph/graph.go (hunk 1 above)
Comment (kellrott) [reply kvgraph/graph.go]: Fixed

File: server/api.go (hunk 2 above)
Comment (kellrott) [reply server/api.go]: The error is copied out now. Note, the current BulkAdd behavior is to cache an error message, keep on processing messages and return the error at the end

File: kvgraph/graph.go [hunk 5]
```diff
@@ -74,47 +57,91 @@ func (kgdb *KVInterfaceGDB) AddVertex(vertices []*gripql.Vertex) error {
...
-			if err != nil {
-				return err
-			}
-
-			src := edge.From
-			dst := edge.To
-			ekey := EdgeKey(kgdb.graph, eid, src, dst, edge.Label, edgeSingle)
-			skey := SrcEdgeKey(kgdb.graph, src, dst, eid, edge.Label, edgeSingle)
-			dkey := DstEdgeKey(kgdb.graph, src, dst, eid, edge.Label, edgeSingle)
+			insertEdge(tx, kgdb.kvg.idx, kgdb.graph, edge)
```
Comment (adamstruck) [inline kvgraph/graph.go]: This error needs to be checked.

File: util/insert.go (hunk 4 above)
Comment (adamstruck) [reply util/insert.go]: Please add logic to ensure each `GraphElement` references the correct graph and add a unit test.

File: kvgraph/graph.go [hunk 6]
```diff
@@ -74,47 +57,91 @@ func (kgdb *KVInterfaceGDB) AddVertex(vertices []*gripql.Vertex) error {
...
-				return err
-			}
-			err = tx.Set(skey, []byte{})
-			if err != nil {
-				return err
-			}
-			err = tx.Set(dkey, []byte{})
-			if err != nil {
-				return err
+func (kgdb *KVInterfaceGDB) BulkAdd(stream <-chan *gripql.GraphElement) error {
```
Comment (adamstruck) [inline kvgraph/graph.go]: This method is missing the validate call on the vertices/edges. 

Can you add unit tests for BulkAdd, AddVertex, and AddEdge that attempts to load a invalid entity.

File: kvgraph/graph.go (hunk 5 above)
Comment (kellrott) [reply kvgraph/graph.go]: Fixed

File: util/insert.go (hunk 4 above)
Comment (kellrott) [reply util/insert.go]: Done.

**Comment by adamstruck** [comment]
I tried loading BMEG with this branch and got the following error:

```
{"level":"info","msg":"Loading outputs/tcga/HNSC.TranscriptExpression.Vertex.json.gz","time":"2019-09-11T22:54:17Z"}
{"level":"info","msg":"Loaded 566 vertices","time":"2019-09-11T22:55:03Z"}
2019/09/11 22:59:03 Assert failed
github.com/dgraph-io/badger/y.AssertTrue
	/go/pkg/mod/github.com/dgraph-io/badger@v2.0.0-rc2+incompatible/y/error.go:55
github.com/dgraph-io/badger.(*DB).updateHead
	/go/pkg/mod/github.com/dgraph-io/badger@v2.0.0-rc2+incompatible/db.go:560
github.com/dgraph-io/badger.(*DB).writeRequests
	/go/pkg/mod/github.com/dgraph-io/badger@v2.0.0-rc2+incompatible/db.go:653
github.com/dgraph-io/badger.(*DB).doWrites.func1
	/go/pkg/mod/github.com/dgraph-io/badger@v2.0.0-rc2+incompatible/db.go:692
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1337
Command exited with non-zero status 1
```

File: cmd/kvload/main.go [hunk 7]
```diff
@@ -87,69 +90,41 @@ var Cmd = &cobra.Command{
 			edgeFileArray = append(edgeFileArray, edgeFile)
 		}
 
+		graphChan := make(chan *gripql.GraphElement, 10)
+		wg := &sync.WaitGroup{}
+		go func() {
+			wg.Add(1)
+			kgraph.BulkAdd(graphChan)
```
[the remark this answers is not in the document]
Comment (kellrott) [reply cmd/kvload/main.go]: Fixed

File: server/api.go [hunk 8]
```diff
@@ -169,137 +171,80 @@ func (server *GripServer) addEdge(ctx context.Context, elem *gripql.GraphElement
...
+						loopErr = err
+						errorCount++
+					}
+					elementStream = make(chan *gripql.GraphElement, 100)
+					wg.Add(1)
+					go func() {
+						log.Printf("Streaming to %s", element.Graph)
+						err := graph.BulkAdd(elementStream)
+						if err != nil {
+							errorCount++
```
[the remark this answers is not in the document]
Comment (kellrott) [reply server/api.go]: Fixed

File: kvgraph/graph.go [hunk 9]
```diff
@@ -74,47 +57,102 @@ func (kgdb *KVInterfaceGDB) AddVertex(vertices []*gripql.Vertex) error {
...
+					if err := insertEdge(tx, kgdb.kvg.idx, kgdb.graph, elem.Edge); err != nil {
+						anyErr = err
+					}
+				} else {
+					anyErr = err
+				}
 			}
 		}
-		kgdb.kvg.ts.Touch(kgdb.graph)
 		return nil
```
Comment (adamstruck) [inline kvgraph/graph.go]: anyErr should be returned here.

**Comment by adamstruck** [comment]
I think we should use something like https://github.com/hashicorp/go-multierror to collect all errors during bulk loading, that way we can return (or log) all errors that were encountered

**Comment by adamstruck** [comment]
Replaced by #223

**Pull request closed without merging**